site/app/controllers/grading/ImagesController.php

```diff
@@ -48,4 +54,139 @@ public function viewImagesPage() {
         $students = $this->core->getQueries()->getAllUsers();
         $this->core->getOutput()->renderOutput(array('grading', 'Images'), 'listStudentImages', $students, $grader_sections, $instructor_permission);
     }
+
+    /**
+     * @Route("/{_semester}/{_course}/student_photos/upload")
+     */
+    public function ajaxUploadImagesFiles() {
+        if(!$this->core->getUser()->accessAdmin()) {
+            return $this->core->getOutput()->renderResultMessage("You have no permission to access this page", false);
+        }
+
+        if (empty($_POST)) {
+            $max_size = ini_get('post_max_size');
+            return $this->core->getOutput()->renderResultMessage("Empty POST request. This may mean that the sum size of your files are greater than {$max_size}.", false, false);
+        }
+
+        if (!isset($_POST['csrf_token']) || !$this->core->checkCsrfToken($_POST['csrf_token'])) {
+            return $this->core->getOutput()->renderResultMessage("Invalid CSRF token.", false, false);
+        }
+
+        $uploaded_files = array();
+        if (isset($_FILES["files1"])) {
+            $uploaded_files[1] = $_FILES["files1"];
+        }
+        $errors = array();
+        $count_item = 0;
+        if (isset($uploaded_files[1])) {
+            $count_item = count($uploaded_files[1]["name"]);
+            for ($j = 0; $j < $count_item[1]; $j++) {
+                if (!isset($uploaded_files[1]["tmp_name"][$j]) || $uploaded_files[1]["tmp_name"][$j] === "") {
+                    $error_message = $uploaded_files[1]["name"][$j]." failed to upload. ";
+                    if (isset($uploaded_files[1]["error"][$j])) {
+                        $error_message .= "Error message: ". ErrorMessages::uploadErrors($uploaded_files[1]["error"][$j]). ".";
+                    }
+                    $errors[] = $error_message;
+                }
+            }
+        }
+
+        if (count($errors) > 0) {
+            $error_text = implode("\n", $errors);
```

Comment:
If this is getting rendered on the client side, probably `\n` should be made into `<br>`

Reply:
Good catch, but this message is also part of JSON. So `\n` is more proper here. I would add an `nl2br` on the twig side. Could you have a look at #4006?